db-service/lib/SQLService.js

```diff
@@ -168,13 +168,21 @@ class SQLService extends DatabaseService {
             where: filtered.where ? [{ xpr: filtered.where }, 'and', { xpr: where }] : where
           })
         }
+
         // Process child compositions depth-first
-        await Promise.all (Object.values(compositions).map(c => {
+        for(const c of Object.values(compositions)) {
           if (c._target['@cds.persistence.skip'] === true) return
-          if (c._target === req.target) if (++depth > (c['@depth'] || 3)) return
-          let query = DELETE.from ({ref:[ ...from.ref, c.name ]}) // CQL`DELETE from Foo[pred]:comp1.comp2...`
-          return this.onDELETE({ query, depth, target: c._target })
-        }))
+          let count = 0
+          for (let i = 0; i < depth.length; i++) {
+            if (depth[i] === c) count++
```

Comment:
revert to hierarchies and throw in case of circle detected.

Reply:
Sollte jetzt passen